### PR TITLE
Update offline sample data handling

### PR DESCRIPTION
**Pull request by tanneryould**
# Description

This PR introduces changes to individual C++ and QML samples as well as to the shared sample viewer code and affects how they store offline data. This was prompted by no longer being able to use Android's `"/sdcard"` root directory. This also provided an opportunity to better utilize the `QStandardPaths` library and standardize how and where we store offline data. Unfortunately, there is a bug with iOS and `QStandardPaths::writablePath(QStandardPaths::HomeLocation)` does not return a writable location so I had to keep iOS using `QStandardPaths::DocumentsLocation`. It'd be great if we could just use a writable home location for all platforms, but this is a serviceable solution for the time being.

## Type of change

- [x] Bug fix
- [x] Sample viewer enhancement
- [x] Other enhancement

**Platforms tested on**:

- [ ] Windows
- [x] Android
- [ ] Linux
- [x] macOS
- [x] iOS

## Checklist

- [x] Runs and compiles on all active platforms as a standalone sample
- [x] Runs and compiles in the sample viewer(s)
- [x] Branch is up to date with the latest main/v.next
- [x] All merge conflicts have been resolved
- [x] Self-review of changes
- [x] There are no warnings related to changes
- [x] No unrelated changes have been made to any other code or project files
- [x] Code is commented with correct formatting (CTRL+i)
- [x] All variable and method names are camel case
- [x] There is no leftover commented code

### CODE REVIEW

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyRasterCell/IdentifyRasterCell.cpp [hunk 1]
```diff
@@ -54,12 +48,10 @@ QString defaultDataPath()
...
   QString dataPath;
 
-#ifdef Q_OS_ANDROID
-  dataPath = "/sdcard";
-#elif defined Q_OS_IOS
+#ifdef Q_OS_IOS
   dataPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
 #else
-  dataPath = QDir::homePath();
+  dataPath = QStandardPaths::writableLocation(QStandardPaths::HomeLocation);
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyRasterCell/IdentifyRasterCell.cpp:54]: If I recall, we did want to keep all the individual logic in each sample so they are all self-contained. On the other hand, a helper would remove a lot of ifdefs, includes and platform boilerplate from many samples.

At this point, the data path is an implementation detail and not a critical part of the ArcGIS Runtime workflow we need to advertise this way. It's worth a conversation.
@ldanzinger what are your thoughts?

File: ArcGISRuntimeSDKQt_QMLSamples/DisplayInformation/ReadSymbolsFromMobileStyle/ReadSymbolsFromMobileStyle.qml [hunk 2]
```diff
@@ -26,7 +26,7 @@ Rectangle {
     width: 800
     height: 600
 
-    readonly property url styleDataPath: System.userHomePath + "/ArcGIS/Runtime/Data/styles/emoji-mobile.stylx"
+    readonly property url styleDataPath: System.writableLocationUrl(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data/styles/emoji-mobile.stylx"
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/DisplayInformation/ReadSymbolsFromMobileStyle/ReadSymbolsFromMobileStyle.qml:29]: Does this need conditional logic for iOS (`StandardPathsDocumentsLocation`)?

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/AddEncExchangeSet/AddEncExchangeSet.qml [hunk 3]
```diff
@@ -24,7 +24,7 @@ Rectangle {
     width: 800
     height: 600
 
-    property url dataPath: System.userHomePath + "/ArcGIS/Runtime/Data"
+    property url dataPath: System.writableLocationUrl(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data"
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/AddEncExchangeSet/AddEncExchangeSet.qml:27]: Ditto conditionalize for iOS?

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/DisplayKml/DisplayKml.qml [hunk 4]
```diff
@@ -139,7 +139,7 @@ Rectangle {
     function createFromFile() {
         // create the dataset from a local file
         const kmlDataset = ArcGISRuntimeEnvironment.createObject("KmlDataset", {
-                                                                     url: System.userHomePath + "/ArcGIS/Runtime/Data/kml/US_State_Capitals.kml"
+                                                                     url: System.writableLocationUrl(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data/kml/US_State_Capitals.kml"
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/DisplayKml/DisplayKml.qml:142]: Does this need an iOS conditional?

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/RasterFunctionFile/RasterFunctionFile.qml [hunk 5]
```diff
@@ -25,7 +25,7 @@ Rectangle {
     height: 600
 
     property real scaleFactor: 1
-    property url dataPath: System.userHomePath + "/ArcGIS/Runtime/Data/raster"
+    property url dataPath: System.writableLocationUrl(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data/raster"
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/RasterFunctionFile/RasterFunctionFile.qml:28]: Ditto iOS conditional?

Since there are a few of these, I won't comment on them all.

File: ArcGISRuntimeSDKQt_QMLSamples/Scenes/AnimateImagesWithImageOverlay/AnimateImagesWithImageOverlay.qml [hunk 6]
```diff
@@ -28,7 +28,11 @@ Rectangle {
 
     readonly property var timerIntervals: [17,34,68]
     readonly property var imageFrameRefreshRate: ["Fast","Medium","Slow"]
-    readonly property url dataPath: System.userHomePath +  "/ArcGIS/Runtime/Data/3D/ImageOverlay/PacificSouthWest"
+    readonly property url dataPath: {
+        Qt.platform.os === "ios" ?
+                    System.writableLocationUrl(System.StandardPathsDocumentsLocation) + "/ArcGIS/Runtime/Data/3D/ImageOverlay/PacificSouthWest" :
+                    System.writableLocationUrl(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data/3D/ImageOverlay/PacificSouthWest"
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_QMLSamples/Scenes/AnimateImagesWithImageOverlay/AnimateImagesWithImageOverlay.qml:34]: I don't think the qml ternaries are on the same level of boilerplate as C++, so I would not argue they need a common helper.

File: ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/ArcGISRuntimeSDKQt_CppSamples.pro [hunk 7]
```diff
@@ -182,8 +180,6 @@ mac {
 }
 
 android {
-    QT += androidextras
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/ArcGISRuntimeSDKQt_CppSamples.pro:185]: Wow I'm surprised this survived our Qt 6 builds this long. Good find.

File: ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml [hunk 8]
```diff
@@ -11,73 +12,17 @@
...
-            <!-- extract android style -->
-            <!-- available android:values :
-                * default - In most cases this will be the same as "full", but it can also be something else if needed, e.g., for compatibility reasons
-                * full - useful QWidget & Quick Controls 1 apps
-                * minimal - useful for Quick Controls 2 apps, it is much faster than "full"
-                * none - useful for apps that don't use any of the above Qt modules
-                -->
+            <meta-data android:name="android.app.arguments" android:value="-- %%INSERT_APP_ARGUMENTS%% --"/>
             <meta-data android:name="android.app.extract_android_style" android:value="default"/>
-            <!-- extract android style -->
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml:78]: Can you summarize the changes to the manifest files? I didn't expect to see those.

File: ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_Samples/qml/ManageOfflineDataView.qml [hunk 9]
```diff
@@ -72,7 +72,14 @@ Page {
...
             Layout.alignment: Qt.AlignHCenter | Qt.AlignVCenter
             visible: !SampleManager.downloadInProgress
-            onClicked: messageDialog.visible = fileFolder.removeFolder(System.userHomePath + "/ArcGIS/Runtime/Data", true);
+            onClicked: {
+                let folderPath;
+                if (Qt.platform.os === "ios")
+                    folderPath = System.writableLocation(System.StandardPathsDocumentsLocation) + "/ArcGIS/Runtime/Data";
+                else
+                    folderPath = System.writableLocation(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data";
+                messageDialog.visible = fileFolder.removeFolder(folderPath, true);
```
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_Samples/qml/ManageOfflineDataView.qml:81]:
```suggestion
                    folderPath = System.writableLocation(System.StandardPathsHomeLocation) + "/ArcGIS/Runtime/Data";

                messageDialog.visible = fileFolder.removeFolder(folderPath, true);
```

File: ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml (hunk 8 above)
Comment (tanneryould) [reply ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml:78]: In short, we had to regenerate the Android Manifests because the `Ministro` service was discontinued and that was causing the compiler to crash. These new Android Manifests represent the default Qt 6 version.

### [Qt Manifest before 6.2 Release](https://doc.qt.io/qt-6.2/android-manifest-file-configuration.html#qt-manifest-before-6-2-release)

Versions of Qt earlier than 6.2 used to have an additional set of meta-data defined by Qt. These attributes used to manage dependencies and some were used by the discontinued Ministro service. With Qt 6.2, they should be removed. Here is a list of these attributes:

- android.app.qt_sources_resource_id
- android.app.repository
- android.app.bundled_libs_resource_id
- android.app.bundle_local_qt_libs
- android.app.use_local_qt_libs
- android.app.libs_prefix
- android.app.load_local_libs_resource_id
- android.app.load_local_jars
- android.app.static_init_classes
- android.app.qt_libs_resource_id
- android.app.ministro_not_found_msg
- android.app.ministro_needed_msg
- android.app.fatal_error_msg

File: ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml [hunk 10]
```diff
@@ -11,73 +12,17 @@
...
-
-    <supports-screens android:largeScreens="true" android:normalScreens="true" android:anyDensity="true" android:smallScreens="true"/>
-
-    <application android:hardwareAccelerated="true" android:name="org.qtproject.qt5.android.bindings.QtApplication" android:label="Qt Samples - C++" android:icon="@drawable/icon" android:extractNativeLibs="true">
-        <activity android:configChanges="locale|fontScale|keyboard|keyboardHidden|layoutDirection|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode" android:name="org.qtproject.qt5.android.bindings.QtActivity" android:label="Qt Samples - C++" android:screenOrientation="unspecified" android:launchMode="singleTop">
+    <supports-screens android:anyDensity="true" android:largeScreens="true" android:normalScreens="true" android:smallScreens="true"/>
+    <application android:name="org.qtproject.qt.android.bindings.QtApplication" android:hardwareAccelerated="true" android:label="Qt Samples - C++" android:requestLegacyExternalStorage="true" android:allowNativeHeapPointerTagging="false" android:allowBackup="true" android:fullBackupOnly="false" android:icon="@drawable/icon">
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml:17]: > android:requestLegacyExternalStorage

Do we still need this? For the time being, I think it's fine if we do. But I was hoping not...

Comment (tanneryould) [reply ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/android/AndroidManifest.xml:17]: I can verify if we need that or not. I believe that was a part of the base template though.

**Comment by tanneryould** [comment]
Thanks for your feedback. I believe I've addressed everything found thus far and this is ready for a re-review

**Review by ldanzinger — APPROVED** [review]

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyRasterCell/IdentifyRasterCell.cpp (hunk 1 above)
Comment (JamesMBallard) [reply ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyRasterCell/IdentifyRasterCell.cpp:54]: I'll open a separate issue for the discussion. Nothing required here.

**Review by JamesMBallard — APPROVED** [review]
